shared/src/state-migrations/migrations.ts

```diff
@@ -1,76 +1,167 @@
-        if (actionMigration !== null) {
-            actionMigration(
-                propertiesToMigrate.history.initialState,
-                propertiesToMigrate.history.actions
-            );
+): {
+    newVersion: number;
+    migratedProperties: {
+        currentState: Mutable<ExerciseState>;
+        history: H extends undefined
+            ? undefined
+            :
+                  | {
+                        initialState: Mutable<ExerciseState>;
+                        actions: (Mutable<ExerciseAction> | null)[];
+                    }
+                  | undefined;
+    };
+} {
+    const newVersion = ExerciseState.currentStateVersion;
+
+    const migrationsToApply: Migration[] = [];
+    for (let i = currentStateVersion + 1; i <= newVersion; i++) {
+        migrationsToApply.push(migrations[i]!);
+    }
+
+    const history = propertiesToMigrate.history;
+    if (history) {
+        migrateState(migrationsToApply, history.initialState);
+        const intermediaryState = cloneDeepMutable(
+            history.initialState
+        ) as Mutable<ExerciseState>;
+        try {
+            history.actions.forEach((action, index) => {
+                if (action !== null) {
+                    const deleteAction = !migrateAction(
+                        migrationsToApply,
+                        intermediaryState,
+                        action
+                    );
+                    if (!deleteAction) {
+                        try {
+                            applyAction(
+                                intermediaryState,
+                                action as ExerciseAction
+                            );
+                        } catch (e: unknown) {
+                            if (e instanceof ReducerError) {
+                                const json = JSON.stringify(action);
+                                console.warn(
+                                    `Error while applying action ${json}: ${e.message}`
+                                );
+                            }
+                            throw e;
+                        }
+                    } else {
+                        history.actions[index] = null;
+                    }
+                }
+            });
+            return {
+                newVersion,
+                migratedProperties: {
+                    currentState: intermediaryState,
+                    // history has been migrated in place
+                    history: history as any,
+                },
+            };
+        } catch (e: unknown) {
+            if (e instanceof ReducerError) {
+                const exerciseId = (history.initialState as { id: UUID }).id;
+                console.warn(
+                    `Discarding history of exercise ${exerciseId} due to error in applying actions: ${e.message}`,
+                    e.stack
+                );
+                // Fall back to migrating currentState instead of recreating it from history
```

Comment:
No, it is in the right place. If no history is present, `currentState` is used as first choice, not as fallback. This comment should express that the function does not terminate immediately in case of the error.

Reply:
Then I would expect it to be immediately above or below the `if (e instanceof ReducerError) {` line, since in its current position I thought it was describing the

```ts
} else {
    throw e;
}
```

part of the code